### ISSUE TITLE
Unhandled General SOCKS Server failure (1)

### ISSUE THREAD

**Issue by NSkelsey** [globaleaks/GlobaLeaks#2111]
**Current behavior**

The exception is generated in exit_nodes_refresh_sched and in 

```
Version: 2.72.7

txsocksx.errors.ServerFailure
    General SOCKS server failure ( 1 )


Traceback (most recent call last):

  File "/usr/lib/python2.7/dist-packages/globaleaks/jobs/base.py", line 78, in run
    yield self.operation()

  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)

  File "/usr/lib/python2.7/dist-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)

  File "/usr/lib/python2.7/dist-packages/globaleaks/jobs/exit_nodes_refresh_sched.py", line 23, in operation
    yield State.tor_exit_set.update(net_agent)

  File "/usr/lib/python2.7/dist-packages/ometa/protocol.py", line 53, in dataReceived
    self._parser.receive(data)

  File "/usr/lib/python2.7/dist-packages/ometa/tube.py", line 41, in receive
    status = self._interp.receive(data)

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 48, in receive
    for x in self.next:

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 166, in apply
    for x in self._apply(f, ruleName, argvals):

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 104, in _apply
    for x in rule():

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 245, in parse_Or
    for x in self._eval(subexpr):

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 230, in parse_And
    for x in self._eval(subexpr):

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 423, in parse_Action
    val = eval(expr.data, self.globals, self._localsStack[-1])

  File "<string>", line 1, in <module>

  File "/usr/lib/python2.7/dist-packages/txsocksx/client.py", line 153, in serverResponse
    raise e.socks5ErrorMap.get(status)()

ServerFailure
```

And inside of notification_sched
```
Version: 2.72.7

txsocksx.errors.ServerFailure
    General SOCKS server failure ( 1 )


Traceback (most recent call last):

  File "/usr/lib/python2.7/dist-packages/globaleaks/jobs/base.py", line 76, in run
    yield threads.deferToThread(self.operation)

  File "/usr/lib/python2.7/dist-packages/twisted/python/threadpool.py", line 246, in inContext
    result = inContext.theWork()

  File "/usr/lib/python2.7/dist-packages/twisted/python/threadpool.py", line 262, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)

  File "/usr/lib/python2.7/dist-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)

  File "/usr/lib/python2.7/dist-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)

  File "/usr/lib/python2.7/dist-packages/globaleaks/jobs/notification_sched.py", line 266, in operation
    self.spool_emails()

  File "/usr/lib/python2.7/dist-packages/globaleaks/jobs/notification_sched.py", line 261, in spool_emails
    threads.blockingCallFromThread(reactor, self.sendmail, mail)

  File "/usr/lib/python2.7/dist-packages/twisted/internet/threads.py", line 122, in blockingCallFromThread
    result.raiseException()

  File "/usr/lib/python2.7/dist-packages/ometa/protocol.py", line 53, in dataReceived
    self._parser.receive(data)

  File "/usr/lib/python2.7/dist-packages/ometa/tube.py", line 41, in receive
    status = self._interp.receive(data)

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 48, in receive
    for x in self.next:

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 166, in apply
    for x in self._apply(f, ruleName, argvals):

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 104, in _apply
    for x in rule():

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 245, in parse_Or
    for x in self._eval(subexpr):

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 230, in parse_And
    for x in self._eval(subexpr):

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 423, in parse_Action
    val = eval(expr.data, self.globals, self._localsStack[-1])

  File "<string>", line 1, in <module>

  File "/usr/lib/python2.7/dist-packages/txsocksx/client.py", line 153, in serverResponse
    raise e.socks5ErrorMap.get(status)()

ServerFailure
```

**Expected behavior**

The exception is not thrown or the issue is understood and handled.

**Steps to reproduce the problem or feature illustration**

Have the Tor SOCKS server throw an exception whose meaning can be found via RFC.

**GlobaLeaks version:**

v2.72.7 and below.

**Comment by NSkelsey** [comment]
Also in `tor_exit_set`

```
Version: 2.71.2

txsocksx.errors.ServerFailure
    General SOCKS server failure ( 1 )


Traceback (most recent call last):

  File "/usr/lib/python2.7/dist-packages/globaleaks/jobs/base.py", line 45, in run
    yield self.operation()

  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)

  File "/usr/lib/python2.7/dist-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)

  File "/usr/lib/python2.7/dist-packages/globaleaks/jobs/exit_nodes_refresh_sched.py", line 23, in operation
    yield GLSettings.appstate.tor_exit_set.update(net_agent)

  File "/usr/lib/python2.7/dist-packages/ometa/protocol.py", line 53, in dataReceived
    self._parser.receive(data)

  File "/usr/lib/python2.7/dist-packages/ometa/tube.py", line 41, in receive
    status = self._interp.receive(data)

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 48, in receive
    for x in self.next:

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 166, in apply
    for x in self._apply(f, ruleName, argvals):

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 104, in _apply
    for x in rule():

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 245, in parse_Or
    for x in self._eval(subexpr):

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 230, in parse_And
    for x in self._eval(subexpr):

  File "/usr/lib/python2.7/dist-packages/ometa/interp.py", line 423, in parse_Action
    val = eval(expr.data, self.globals, self._localsStack[-1])

  File "<string>", line 1, in <module>

  File "/usr/lib/python2.7/dist-packages/txsocksx/client.py", line 153, in serverResponse
    raise e.socks5ErrorMap.get(status)()

ServerFailure
```

**Comment by NSkelsey** [comment]
Looking at these exception in the notification_sched job yields an explanation as to why these failures are reaching the dev mailing list, despite the tight exception handling around `utils.mailutils.sendmail`

in `utils.mailutils.sendmail` we have
```python
try:
  # lots of actions
  # read db 
  # blah blah
  d.endpoint.connection(factory)
  d.adderrback(errback)
except Exception as e:
  log(e)
```

And in notification_seched we have the usage of a [threaded call](http://twistedmatrix.com/documents/current/api/twisted.internet.threads.html#blockingCallFromThread):

```python
for mail in get_mails_from_the_pool():
   threads.blockingCallFromThread(reactor, self.sendmail, mail)
```

This function waits for the deferred to execute and rethrows any fails encountered in the deferred as regular exceptions. Since the networking failures are the only ones that regularly occur in the deferred, they are not eaten by the try catch block and we continue to get emails about them.

It seems safe to wrap the notification_sched as an ExternNetLoopingJob, to catch these specific failures. Further the behavior in my opinion seems correct, where one networking failure on the spool of mails kills the for loop. I do not see a point retrying the rest of the mails when the endpoint.connection(factory) method has failed. It seems better to just back off wait for the job to retry at the next run of the job.

**Comment by NSkelsey** [comment]
Should be resolved in 28febc0
